### PR TITLE
feat(STONEINTG-999): report test status for group snapshot

### PR DESCRIPTION
**Pull request by hongweiliu17**
* update template to generate text reported to git provider
* report test status for group snapshot

Signed-off-by: Hongwei Liu <hongliu@redhat.com>

## Maintainers will complete the following section

- [ ] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [ ] Code coverage from testing does not decrease and new code is covered ([check the PR coverage on codecov](https://app.codecov.io/gh/redhat-appstudio/integration-service/pulls))
- [ ] [Controllers diagrams](https://github.com/konflux-ci/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

**Comment by hongweiliu17** [comment]
The new text with component snapshot info for group snapshot is like below:  <li><b>Pipelinerun</b>: <a href="https://definetly.not.prod/preview/application-pipeline/ns/default/pipelinerun/pipelinerun-component-sample">pipelinerun-component-sample</a></li>
  </ul>
  <hr>

  | Task | Duration | Test Suite | Status | Details |
  | --- | --- | --- | --- | --- |
  | <a href="https://definetly.not.prod/preview/application-pipeline/ns/default/pipelinerun/pipelinerun-component-sample/logs/example-task-1">example-task-1</a> | 5m30s |  | :heavy_check_mark: Reason: Succeeded | :heavy_check_mark: Reason: Succeeded |



  The group snapshot is generated for the component snasphots as below:
  | Component | Snapshot | BuildPipelineRun |
  | --- | --- | --- |
  | com1 | snapshot1 | <a href="https://definetly.not.prod/preview/application-pipeline/ns/default/pipelinerun/buildPLR1">buildPLR1</a> |
  | com2 | snapshot2 | <a href="https://definetly.not.prod/preview/application-pipeline/ns/default/pipelinerun/buildPLR2">buildPLR2</a> |
  | com3 | snapshot3 | <a href="https://definetly.not.prod/preview/application-pipeline/ns/default/pipelinerun/buildPLR3">buildPLR3</a> |

**Comment by hongweiliu17** [comment]
I have found a lot of wrong code in this PR, so please don't spend too much time to review it, thanks!

File: status/status.go [hunk 1]
```diff
@@ -533,3 +576,30 @@ func (s Status) IsPRInSnapshotOpened(ctx context.Context, reporter ReporterInter
...
+	}
+
+	for _, componentSnapshotInfo := range componentSnapshotInfos {
+		componentSnapshot := &applicationapiv1alpha1.Snapshot{}
+		err := c.Get(ctx, types.NamespacedName{
+			Namespace: groupSnapshot.Namespace,
+			Name:      componentSnapshotInfo.Snapshot,
+		}, componentSnapshot)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find component snapshot %s included in group snapshot %s/%s: %w", componentSnapshotInfo.Snapshot, groupSnapshot.Namespace, groupSnapshot.Name, err)
```
Comment (dirgim) [inline status/status.go:599]: I'm thinking if we should log the error here but continue the loop anyways to ensure that the other Snapshot PRs get the report sent to them? I'm not sure if that's the way to go, wdyt?

Comment (hongweiliu17) [reply status/status.go:599]: It makes sense, let me update it. Thanks.

Comment (hongweiliu17) [reply status/status.go]: Updated.

File: internal/controller/statusreport/statusreport_adapter.go [hunk 2]
```diff
@@ -66,19 +68,14 @@ func NewAdapter(context context.Context, snapshot *applicationapiv1alpha1.Snapsh
 
 // EnsureSnapshotTestStatusReportedToGitProvider will ensure that integration test status is reported to the git provider
 // which (indirectly) triggered its execution.
+// The status is reported to git provider if it is a component snapshot
+// Or reported to git providers which trigger component snapshots included in group snapshot if it is a gorup snapshot
```
Comment (dirgim) [inline internal/controller/statusreport/statusreport_adapter.go:72]: tiny typo: `gorup snapshot` -> `group snapshot`

File: internal/controller/statusreport/statusreport_adapter.go [hunk 3]
```diff
@@ -238,3 +235,108 @@ func (a *Adapter) findUntriggeredIntegrationTestFromStatus(integrationTestScenar
...
+		destinationSnapshots = append(destinationSnapshots, testedSnapshot)
+	} else if gitops.IsGroupSnapshot(testedSnapshot) {
+		// get component snapshots from group snapshot annation GroupSnapshotInfoAnnotation
+		destinationSnapshots, err = status.GetComponentSnapshotsFromGroupSnapshot(ctx, a.client, testedSnapshot)
+		if err != nil {
+			a.logger.Error(err, "failed to get component snapshots included in group snapshot",
+				"snapshot.NameSpace", testedSnapshot.Namespace, "snapshot.Name", testedSnapshot.Name)
+			return fmt.Errorf("failed to get component snapshots included in group snapshot %s/%s", testedSnapshot.Namespace, testedSnapshot.Name)
+		}
+	}
```
Comment (dirgim) [inline internal/controller/statusreport/statusreport_adapter.go:269]: Maybe we can extract this logic for getting destination Snapshots to a separate function to reduce the cognitive complexity for `ReportSnapshotStatus` (currently 20).

File: internal/controller/statusreport/statusreport_adapter.go [hunk 4]
```diff
@@ -238,3 +235,108 @@ func (a *Adapter) findUntriggeredIntegrationTestFromStatus(integrationTestScenar
...
+					}
+					return fmt.Errorf("failed to generate test report: %w", reportErr)
+				}
+				if reportStatusErr := reporter.ReportStatus(ctx, *testReport); reportStatusErr != nil {
+					if writeErr := status.WriteSnapshotReportStatus(ctx, a.client, testedSnapshot, srs); writeErr != nil { // try to write what was already written
+						return fmt.Errorf("failed to report status AND write snapshot report status metadata: %w", e.Join(reportStatusErr, writeErr))
+					}
+					return fmt.Errorf("failed to update status: %w", reportStatusErr)
+				}
+				srs.SetLastUpdateTime(integrationTestStatusDetail.ScenarioName, integrationTestStatusDetail.LastUpdateTime)
```
Comment (dirgim) [inline internal/controller/statusreport/statusreport_adapter.go:326]: Maybe we can extract this logic to a separate function that creates and updates the integration report status to reduce the cognitive complexity for `ReportSnapshotStatus` (currently 20).

File: internal/controller/statusreport/statusreport_adapter.go [hunk 5]
```diff
@@ -238,3 +235,108 @@ func (a *Adapter) findUntriggeredIntegrationTestFromStatus(integrationTestScenar
...
+		return nil
+	}
+
+	// get the component snapshot list that include the git provider info the report will be reported to
+	destinationSnapshots := make([]*applicationapiv1alpha1.Snapshot, 0)
+	if gitops.IsComponentSnapshot(testedSnapshot) {
+		destinationSnapshots = append(destinationSnapshots, testedSnapshot)
+	} else if gitops.IsGroupSnapshot(testedSnapshot) {
+		// get component snapshots from group snapshot annation GroupSnapshotInfoAnnotation
+		destinationSnapshots, err = status.GetComponentSnapshotsFromGroupSnapshot(ctx, a.client, testedSnapshot)
```
Comment (dirgim) [inline internal/controller/statusreport/statusreport_adapter.go:263]: tiny typo: `snapshot annation` -> `snapshot annotation`

Comment (hongweiliu17) [reply internal/controller/statusreport/statusreport_adapter.go:263]: Thanks, updated.

File: internal/controller/statusreport/statusreport_adapter.go (hunk 2 above)
Comment (hongweiliu17) [reply internal/controller/statusreport/statusreport_adapter.go]: Thanks, updated.

File: internal/controller/statusreport/statusreport_adapter.go (hunk 3 above)
Comment (hongweiliu17) [reply internal/controller/statusreport/statusreport_adapter.go]: Thanks. I will work on it.

File: internal/controller/statusreport/statusreport_adapter.go (hunk 4 above)
Comment (hongweiliu17) [reply internal/controller/statusreport/statusreport_adapter.go:326]: Thanks. I will work on it.

Comment (hongweiliu17) [reply internal/controller/statusreport/statusreport_adapter.go]: Updated.

File: internal/controller/statusreport/statusreport_adapter.go (hunk 3 above)
Comment (hongweiliu17) [reply internal/controller/statusreport/statusreport_adapter.go]: Updated.

**Review by dirgim — APPROVED** [review]
LGTM

**Review by chipspeak — APPROVED** [review]
LGTM Inspirational stuff as always Hongwei!

**Comment by hongweiliu17** [comment]
Test on my personal cluster and it can create checkRun successfully, refer to https://github.com/hongweiliu17/devfile-sample-go-basic-1772/pull/1/checks?check_run_id=30507639589 and https://github.com/hongweiliu17/devfile-sample-java-springboot-basic-1772/pull/1/checks?check_run_id=30507642618

**Comment by hongweiliu17** [comment]
Found one bug in when [SetLastUpdateTime](https://github.com/konflux-ci/integration-service/pull/860/files#diff-90657e2d0ed009d9e0ef608210c72302bce4bd46e1d995e367442d60d9ebe9bcR357), because this is a loop, so we may need to also include the `componentName` in https://github.com/konflux-ci/integration-service/blob/main/status/status.go#L48. To make the review easier, I prefer to merge this PR then commit another PR to fix it.

**Pull request merged**